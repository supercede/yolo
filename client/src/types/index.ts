export type Event = {
  type: string;
  payload: Payload;
};

export type Payload = {
  id: string;
  competitors: Competitor[];
  markets: Market[];
  startTime: Date;
  updatedAt: Date;
};

export type Competitor = {
  id: string;
  name: string;
  score: number;
};

export type Market = {
  id: string;
  name: string;
  selections: Selection[];
};

export type Selection = {
  id: string;
  name: string;
  odds: number | null;
};

export type EventContainerProps = {
  event: Payload;
};

export type EventTeamProps = {
  competitor: {
    score: number;
    name: string;
  };
};

export type EventMarketProps = {
  markets?: Market[];
};

export type DateTimeProps = {
  startTime: Date;
};
