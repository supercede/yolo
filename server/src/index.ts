import 'module-alias/register';
import server from '@app/api/server';
import { logger } from '@app/api/utils/logger';
import { config } from '@app/config';

void server()
  .then((httpServer) => {
    httpServer.listen({ port: config.server.port }, () => {
      logger.info(
        `⚡️[server]: Server is running at http://localhost:${config.server.port}/graphql`
      );
    });
  })
  .catch((err) => logger.fatal({ err }, 'bootstraping error'));
