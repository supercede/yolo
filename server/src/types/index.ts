type Category = {
  id: string;
  slug: string;
  translations: Translation[];
};

type Translation = {
  id?: string;
  lang: string;
  value: string;
};

type Competitor = {
  id: string;
  name: string;
  score: number;
  translations: Translation[];
};

type Selection = {
  id: string;
  name: string;
  odds: number;
};

type Market = {
  id: string;
  name: string;
  selections: Selection[];
  translations: Translation[];
};

type Payload = {
  id: string;
  category: Category;
  competitors: Competitor[];
  markets: Market[];
  startTime: Date;
  updatedAt: Date;
};

export type Event = {
  type: eventUpdate;
  payload: Payload;
};

enum eventUpdate {
  DATA_EVENT = 'event-data',
  UPDATE_EVENT = 'event-update',
}

export enum Constants {
  UPDATE_EVENT = 'EVENT_UPDATED',
}
