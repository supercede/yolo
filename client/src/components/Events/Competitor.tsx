import React from "react";
import { EventTeamProps } from "src/types";

export function EventCompetitor(props: EventTeamProps) {
  return (
    <li className="flex justify-between mt-1">
      <p className="text-xs text-white sm:text-sm" data-testid="competitor-name">{props.competitor?.name}</p>
      <p className="text-xs text-green sm:text-sm" data-testid="competitor-score">{props.competitor?.score}</p>
    </li>
  );
}
