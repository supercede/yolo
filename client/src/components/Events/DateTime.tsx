import { DateTimeProps } from "src/types";

export function EventDateTime(props: DateTimeProps) {
  const startDate = new Date(props.startTime).toLocaleDateString("en-GB");
  const startTime = new Date(props.startTime).toLocaleTimeString("en-GB", {
    hour12: false,
  });

  return (
    <p className="text-xs sm:text-sm text-grey">
      <span data-testid="event-start-date" data-test>
        {startDate.substring(0, startDate.length - 5)}{" "}
      </span>
      <span data-testid="event-start-time" data-test>
        {startTime.substring(0, startTime.length - 3)}
      </span>
    </p>
  );
}
