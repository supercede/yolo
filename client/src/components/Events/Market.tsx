import { EventContainerProps, EventMarketProps, Selection } from "src/types";

export function EventMarket(props: EventMarketProps) {
    return (
        <ul className="flex w-1/2 gap-2 p-2">
        {props?.markets?.[0] ? (
          props.markets[0]?.selections?.map((selection: Selection) => (
            <li
              key={selection.id}
              className="flex items-center justify-center w-full h-full text-white rounded bg-bg-blue"
              data-testid={`market-odds-${selection.id}`}
            >
              {selection.odds ? Number(selection.odds)?.toFixed(2) : "-"}
            </li>
          ))
        ) : (
          <div className="flex items-center justify-center w-full h-full text-white rounded bg-bg-blue">
            <p className="text-sm text-grey" data-testid="no-market">No market available</p>
          </div>
        )}
      </ul>
    )
}