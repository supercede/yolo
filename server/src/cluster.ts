import 'module-alias/register';
import os from 'os';
import throng from 'throng';
import server from '@app/api/server';
import { logger } from './api/utils/logger';

const startApplication = (): void => {
  void server().catch((err: Error) =>
    logger.fatal({ err }, 'bootstraping error')
  );
};

const startWorkerProcess = (id: number): void => {
  logger.info(`Started worker ${id.toString()}.`);
  startApplication();
};

const startMasterProcess = (): void => {
  logger.info('Master process started.');
};

throng({
  count: os.cpus().length,
  grace: 3000,
  master: startMasterProcess,
  start: startWorkerProcess,
});
