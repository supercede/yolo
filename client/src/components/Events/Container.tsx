import { EventCompetitor } from "./Competitor";
import React from "react";
import { Competitor, EventContainerProps, Selection } from "src/types";
import { EventMarket } from "./Market";
import { EventDateTime } from "./DateTime";

export function EventContainer(props: EventContainerProps) {
  return (
    <li className="flex rounded-sm bg-blue">
      <div className="flex flex-col w-2/3 p-3">
        <ul className="flex flex-col pr-2 mt-1">
          {props.event.competitors.map((competitor: Competitor) => (
            <EventCompetitor competitor={competitor} key={competitor.id} />
          ))}
        </ul>
        <EventDateTime startTime={props.event.startTime} />
      </div>
      <EventMarket markets={props.event?.markets} />
    </li>
  );
}
