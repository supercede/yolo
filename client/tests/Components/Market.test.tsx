import React from "react";
import '@testing-library/jest-dom/extend-expect'
import { render, screen } from "@testing-library/react";
import { EventMarket } from "@/components/Events/Market";
import data from "../data";
import { Market } from "src/types";

test("Return Market Odds",  () => {
  const {markets} = data.event.payload
  const {container} = render(
    <EventMarket
      markets={markets}
    />
  );
  
  expect(container.firstChild).toMatchSnapshot()
  expect(screen.getByTestId(`market-odds-${markets[0].selections[0].id}`)).toHaveTextContent(markets[0].selections[0].odds.toFixed(2));
});

test("Return 'No Market Available' when no market is available", () => {
    // const markets = ([] as unknown) as Market[]
    const markets = [] as Market[]
    const {container} = render(
      <EventMarket
        markets={markets}
      />
    );
    
    expect(container.firstChild).toMatchSnapshot()
    expect(screen.getByTestId('no-market')).toHaveTextContent('No market available');
  });