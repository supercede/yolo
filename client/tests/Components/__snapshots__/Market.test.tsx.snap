// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Return 'No Market Available' when no market is available 1`] = `
<ul
  class="flex w-1/2 gap-2 p-2"
>
  <div
    class="flex items-center justify-center w-full h-full text-white rounded bg-bg-blue"
  >
    <p
      class="text-sm text-grey"
      data-testid="no-market"
    >
      No market available
    </p>
  </div>
</ul>
`;

exports[`Return Market Odds 1`] = `
<ul
  class="flex w-1/2 gap-2 p-2"
>
  <li
    class="flex items-center justify-center w-full h-full text-white rounded bg-bg-blue"
    data-testid="market-odds-49f594fc-85be-444b-b357-e394964041fd"
  >
    8.70
  </li>
  <li
    class="flex items-center justify-center w-full h-full text-white rounded bg-bg-blue"
    data-testid="market-odds-f88a5308-fcdb-4274-8d53-b8fe687e8144"
  >
    9.18
  </li>
  <li
    class="flex items-center justify-center w-full h-full text-white rounded bg-bg-blue"
    data-testid="market-odds-14a4f223-b1e4-458b-a26b-c351a651daca"
  >
    6.71
  </li>
</ul>
`;
